hbase-server/src/main/java/org/apache/hadoop/hbase/security/access/SnapshotScannerHDFSAclCleaner.java

```diff
@@ -0,0 +1,156 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.security.access;
+
+import java.io.IOException;
+import java.util.Map;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hbase.HBaseInterfaceAudience;
+import org.apache.hadoop.hbase.HConstants;
+import org.apache.hadoop.hbase.MetaTableAccessor;
+import org.apache.hadoop.hbase.TableName;
+import org.apache.hadoop.hbase.coprocessor.CoprocessorHost;
+import org.apache.hadoop.hbase.master.HMaster;
+import org.apache.hadoop.hbase.master.cleaner.BaseHFileCleanerDelegate;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.apache.yetus.audience.InterfaceStability;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.hbase.thirdparty.com.google.common.annotations.VisibleForTesting;
+
+/**
+ * Implementation of a file cleaner that checks if a empty directory with no subdirs and subfiles is
+ * deletable when user scan snapshot feature is enabled
+ */
+@InterfaceAudience.LimitedPrivate(HBaseInterfaceAudience.CONFIG)
+@InterfaceStability.Evolving
+public class SnapshotScannerHDFSAclCleaner extends BaseHFileCleanerDelegate {
+  private static final Logger LOG = LoggerFactory.getLogger(SnapshotScannerHDFSAclCleaner.class);
+
+  private HMaster master;
+  private boolean userScanSnapshotEnabled = false;
+
+  @Override
+  public void init(Map<String, Object> params) {
+    if (params != null && params.containsKey(HMaster.MASTER)) {
+      this.master = (HMaster) params.get(HMaster.MASTER);
+    }
+  }
+
+  @Override
+  public void setConf(Configuration conf) {
+    super.setConf(conf);
+    userScanSnapshotEnabled = isUserScanSnapshotEnabled(conf);
+  }
+
+  @Override
+  protected boolean isFileDeletable(FileStatus fStat) {
+    // This plugin does not handle the file deletions, so return true by default
+    return true;
+  }
+
+  @Override
+  public boolean canDeleteEmptyDirectory(Path dir) {
+    if (userScanSnapshotEnabled) {
+      /*
+       * If user scan snapshot feature is enabled(see HBASE-21995), then when namespace or table
+       * exists, the archive namespace or table directories should not be deleted because the HDFS
+       * acls are set at these directories; the archive data directory should not be deleted because
+       * the HDFS acls of global permission is set at this directory.
+       */
+      return canDeleteEmptyArchiveDirectory(dir);
+    }
+    return true;
+  }
+
+  private boolean isUserScanSnapshotEnabled(Configuration conf) {
+    String masterCoprocessors = conf.get(CoprocessorHost.MASTER_COPROCESSOR_CONF_KEY);
+    if (conf.getBoolean(SnapshotScannerHDFSAclHelper.USER_SCAN_SNAPSHOT_ENABLE, false)
```

Comment:
Reconsider the USER_SCAN_SNAPSHOT_ENABLE name, seems not  a good name, because even if we don't enable the feature, user can still scan the snapshot ? so better to use SYNC_TABLE_ACL_IN_HDFS or some other config keys to express the point is to sync the table granted acl to HDFS acl, it will help to understand this.
BTW, a deprecated config keys is a trouble, so better to give a suitable name in the first definition.

Reply:
Yes, great suggestions, I think there are some configs should be renamed to a more proper name, I will do this work in another issus.